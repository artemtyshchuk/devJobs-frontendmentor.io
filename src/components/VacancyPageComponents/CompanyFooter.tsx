import styles from "./VacancyPage.module.scss";
import { motion } from "framer-motion";

interface CompanyFooterProps {
  apply?: string;
  position?: string;
}

export const CompanyFooter = ({ apply, position }: CompanyFooterProps) => {
  const mobileScreen = window.matchMedia("(max-width: 576px)").matches;

  return (
    <div className={styles.companyFooter} data-testid="companyFooter">
      <div className={styles.companyFooter_content}>
        <div>
          <p
            className={styles.vacancyDescription_title}
            style={{ margin: "0", display: mobileScreen ? "none" : "block" }}
            data-testid="vacancyDescription_title"
          >
            {position}
          </p>
          <p
            className={styles.vacancyDescription_description}
            style={{
              margin: "0",
              paddingTop: "12px",
              display: mobileScreen ? "none" : "block",
            }}
          >
            So Digital Inc.
          </p>
        </div>
        <div className={styles.companyFooter_applyButton}>
          <a href={apply} data-testid="applyLink">
            <motion.button
              whileHover={{ scale: 1.1 }}
              whileTap={{ scale: 0.9 }}
              transition={{ type: "spring", stiffness: 400, damping: 10 }}
              className={styles.vacancyDescription_button}
              style={{ marginRight: mobileScreen ? "0" : "46px" }}
              data-testid="applyButton"
            >
              Apply Now
            </motion.button>
          </a>
        </div>
      </div>
    </div>
  );
};
