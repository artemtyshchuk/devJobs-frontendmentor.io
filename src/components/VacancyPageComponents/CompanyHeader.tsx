import styles from "./VacancyPage.module.scss";

interface CompanyHeaderProps {
  logo?: string;
  logoBackground?: string;
  company?: string;
  website?: string;
}

export const CompanyHeader = ({
  logo,
  logoBackground,
  company,
  website,
}: CompanyHeaderProps) => {
  return (
    <div className={styles.companyHeader} data-testid="companyHeader">
      <div
        className={styles.companyHeader_logoContainer}
        style={{ backgroundColor: logoBackground }}
      >
        <img src={logo} alt="company logo" data-testid="companyLogo" />
      </div>
      <div className={styles.companyHeader_companyInfoContainer}>
        <p className={styles.companyHeader_companyName}>{company}</p>
        <p className={styles.companyHeader_companyWebsite}>{website}</p>
      </div>
      <div className={styles.companyHeader_companySiteContainer}>
        <a
          href={website}
          target="_blank"
          rel="noreferrer"
          data-testid="companyWebsite"
        >
          <button
            className={styles.companyHeader_companySiteButton}
            data-testid="headerButton"
          >
            Company Site
          </button>
        </a>
      </div>
    </div>
  );
};
