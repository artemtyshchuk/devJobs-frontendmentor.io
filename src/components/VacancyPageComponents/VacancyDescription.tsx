import styles from "./VacancyPage.module.scss";

interface VacancyDescriptionProps {
  postedAt?: string;
  contract?: string;
  location?: string;
  position?: string;
  apply?: string;
  description?: string;
  requirementsContent?: string;
  requirementsItems?: string[];
  roleContent?: string;
  roleItems?: string[];
}

export const VacancyDescription = ({
  apply,
  contract,
  description,
  location,
  position,
  postedAt,
  requirementsContent,
  requirementsItems,
  roleContent,
  roleItems,
}: VacancyDescriptionProps) => {
  return (
    <div className={styles.vacancyDescription_container}>
      <div className={styles.vacancyDescription_header_wrapper}>
        <div>
          <div className={styles.vacancyDescription_header_subInfo}>
            <p className={styles.vacancyDescription_header_subInfo_text}>
              {postedAt}
            </p>
            <p className={styles.vacancyDescription_header_subInfo_text}>â€¢</p>
            <p className={styles.vacancyDescription_header_subInfo_text}>
              {contract}
            </p>
          </div>
          <p className={styles.vacancyDescription_header_position}>
            {position}
          </p>
          <p className={styles.vacancyDescription_header_location}>
            {location}
          </p>
        </div>

        <div className={styles.vacancyDescription_buttonContainer}>
          <a href={apply}>
            <button className={styles.vacancyDescription_button}>
              Apply Now
            </button>
          </a>
        </div>
      </div>

      <p className={styles.vacancyDescription_description}>{description}</p>

      <p className={styles.vacancyDescription_title}>Requirements</p>
      <p className={styles.vacancyDescription_description}>
        {requirementsContent}
      </p>

      {requirementsItems?.map((item: string, index) => (
        <li
          key={index}
          className={styles.vacancyDescription_description_list}
          style={{ marginTop: "8px" }}
        >
          {item}
        </li>
      ))}

      <p className={styles.vacancyDescription_title}>What you will do</p>
      <p className={styles.vacancyDescription_description}>{roleContent}</p>
      {roleItems?.map((item, index) => (
        <li
          key={index}
          className={styles.vacancyDescription_description_list}
          style={{
            marginTop: "8px",
            listStyleType: "none",
            textIndent: "-30px",
          }}
        >
          <span className={styles.indexNumber}>{index + 1}.</span> {item}
        </li>
      ))}
    </div>
  );
};
