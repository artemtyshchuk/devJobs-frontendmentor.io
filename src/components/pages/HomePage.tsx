import { SearchField } from "components/SearchField";
import { VacanciesList } from "components/VacanciesList";

export const HomePage = () => {
  return (
    <>
      <SearchField />
      <VacanciesList />
    </>
  );
};
