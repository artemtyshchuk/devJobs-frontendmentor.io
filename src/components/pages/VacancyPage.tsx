import styles from "../VacancyPageComponents/VacancyPage.module.scss";
import { VacancyType } from "types";
import { useParams } from "react-router";
import { useEffect, useState } from "react";
import { CompanyHeader } from "../VacancyPageComponents/CompanyHeader";
import { VacancyDescription } from "components/VacancyPageComponents/VacancyDescription";
import { CompanyFooter } from "components/VacancyPageComponents/CompanyFooter";

export const VacancyPage = () => {
  const [vacancy, setVacancy] = useState<VacancyType>();
  const params = useParams<{ vacancyId: string }>();

  const fetchVacancyData = async () => {
    try {
      const res = await fetch("/data.json");
      const dataJson = (await res.json()) as VacancyType[];
      const vacancyId = Number(params.vacancyId);
      const currentJobPage = dataJson.find(
        (vacancyData) => vacancyData.id === vacancyId
      );
      setVacancy(currentJobPage);
    } catch (error) {
      console.error("Error fetching data:", error);
    }
  };

  useEffect(() => {
    fetchVacancyData();
  }, [params.vacancyId]);

  return (
    <div className={styles.vacancyPage}>
      <CompanyHeader
        company={vacancy?.company}
        website={vacancy?.website}
        logo={vacancy?.logo}
        logoBackground={vacancy?.logoBackground}
      />
      <VacancyDescription
        apply={vacancy?.apply}
        contract={vacancy?.contract}
        description={vacancy?.description}
        location={vacancy?.location}
        position={vacancy?.position}
        postedAt={vacancy?.postedAt}
        requirementsContent={vacancy?.requirements.content}
        requirementsItems={vacancy?.requirements.items}
        roleContent={vacancy?.role.content}
        roleItems={vacancy?.role.items}
      />
      <CompanyFooter apply={vacancy?.apply} position={vacancy?.position} />
    </div>
  );
};
