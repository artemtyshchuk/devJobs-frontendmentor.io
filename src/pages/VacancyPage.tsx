import styles from "components/VacancyPageComponents/VacancyPage.module.scss";
import { useNavigate, useParams } from "react-router";
import { CompanyHeader } from "components/VacancyPageComponents/CompanyHeader";
import { VacancyDescription } from "components/VacancyPageComponents/VacancyDescription";
import { CompanyFooter } from "components/VacancyPageComponents/CompanyFooter";
import { useFetch } from "hooks/useFetch";
import { AnimatePresence, motion } from "framer-motion";
import { useState } from "react";

export const VacancyPage = () => {
  const [footerVisible, setFooterVisible] = useState(false);

  const { data: vacancies } = useFetch();
  const params = useParams<{ vacancyId: string }>();

  const vacancy = vacancies.find(
    (vacancyData) => vacancyData.id === Number(params.vacancyId)
  );

  const navigate = useNavigate();

  if (!vacancy) {
    navigate("/not-found");
    return null;
  }

  const headerAnimation = {
    hidden: {
      opacity: 0,
      y: -20,
    },
    visible: {
      opacity: 1,
      y: 0,
      transition: {
        duration: 0.5,
        delay: 0.2,
      },
    },
  };

  const descriptionAnimation = {
    hidden: {
      opacity: 0,
      y: -20,
    },
    visible: {
      opacity: 1,
      y: 0,
      transition: {
        duration: 0.5,
        delay: 0.4,
      },
    },
  };

  const footerAnimation = {
    hidden: {
      opacity: 0,
      transition: {
        duration: 0.3,
      },
    },
    visible: {
      opacity: 1,
      transition: {
        duration: 0.8,
      },
    },
  };

  return (
    <div className={styles.vacancyPage}>
      <motion.div initial="hidden" animate="visible" variants={headerAnimation}>
        <CompanyHeader
          company={vacancy.company}
          website={vacancy.website}
          logo={vacancy.logo}
          logoBackground={vacancy.logoBackground}
        />
      </motion.div>
      <motion.div
        initial="hidden"
        animate="visible"
        variants={descriptionAnimation}
      >
        <VacancyDescription
          apply={vacancy.apply}
          contract={vacancy.contract}
          description={vacancy.description}
          location={vacancy.location}
          position={vacancy.position}
          postedAt={vacancy.postedAt}
          requirementsContent={vacancy.requirements.content}
          requirementsItems={vacancy.requirements.items}
          roleContent={vacancy.role.content}
          roleItems={vacancy.role.items}
          setFooterVisible={setFooterVisible}
        />
      </motion.div>

      <AnimatePresence>
        {footerVisible && (
          <motion.div
            initial="hidden"
            animate="visible"
            exit="hidden"
            variants={footerAnimation}
          >
            <CompanyFooter apply={vacancy.apply} position={vacancy.position} />
          </motion.div>
        )}
      </AnimatePresence>
    </div>
  );
};
