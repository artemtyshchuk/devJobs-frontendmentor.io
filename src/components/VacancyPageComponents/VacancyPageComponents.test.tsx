/* eslint-disable @typescript-eslint/no-unused-vars */
import { render, screen } from "@testing-library/react";
import { VacancyPage } from "../../pages/VacancyPage";
import { CompanyFooter } from "./CompanyFooter";
import { CompanyHeader } from "./CompanyHeader";
import { VacancyDescription } from "./VacancyDescription";
import userEvent from "@testing-library/user-event";

jest.mock("react-hook-form", () => ({
  useForm: () => ({
    register: jest.fn(),
    handleSubmit: (fn: Function) => fn,
  }),
}));

const mockMatchMedia = (matches: boolean) =>
  jest.fn().mockImplementation((query) => ({
    matches,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
  }));

describe("VacancyPage", () => {
  let setFooterVisibleMock: jest.Mock;

  beforeAll(() => {
    const observeMock = jest.fn();
    const unobserveMock = jest.fn();
  });

  beforeEach(() => {
    global.IntersectionObserver = jest.fn(function (
      callback: IntersectionObserverCallback,
      options?: IntersectionObserverInit
    ) {
      return {
        root: options?.root || null,
        rootMargin: options?.rootMargin || "0px",
        thresholds: Array.isArray(options?.threshold)
          ? options?.threshold
          : [options?.threshold ?? 0], // Ensure thresholds is a readonly number array
        observe: jest.fn(),
        unobserve: jest.fn(),
        disconnect: jest.fn(),
        takeRecords: jest.fn(),
        callback,
      };
    }) as unknown as typeof IntersectionObserver;
  });

  afterEach(() => {
    jest.clearAllMocks();

    setFooterVisibleMock = jest.fn();
  });

  const defaultProps = {
    postedAt: "3 days ago",
    contract: "Full-Time",
    location: "Remote",
    position: "Senior Developer",
    apply: "/apply",
    description: "This is the job description.",
    requirementsContent: "You must have experience in React and Node.js.",
    requirementsItems: ["React", "Node.js", "TypeScript"],
    roleContent: "You will work on our product team.",
    roleItems: ["Develop features", "Fix bugs", "Write tests"],
    setFooterVisible: jest.fn(),
  };

  it("should render Company Header correctly", () => {
    render(<CompanyHeader company="company" website="website" />);

    const companyLogo = screen.getByTestId("companyLogo");
    expect(companyLogo).toBeInTheDocument();

    const companyWebsite = screen.getByTestId("companyWebsite");
    expect(companyWebsite).toBeInTheDocument();

    const headerButton = screen.getByTestId("headerButton");
    expect(headerButton).toBeInTheDocument();

    const companyPropsName = screen.getByText("company");
    expect(companyPropsName).toBeInTheDocument();

    const companyPropsWebsite = screen.getByText("website");
    expect(companyPropsWebsite).toBeInTheDocument();

    const companyHeader = screen.getByTestId("companyHeader");
    expect(companyHeader).toMatchSnapshot();
  });
  it("should render Company Footer correctly on desktop", () => {
    window.matchMedia = mockMatchMedia(false);

    render(<CompanyFooter apply="apply" position="position" />);

    const companyFooter = screen.getByTestId("companyFooter");
    expect(companyFooter).toBeInTheDocument();

    const applyButton = screen.getByTestId("applyButton");
    expect(applyButton).toBeInTheDocument();

    const applyLink = screen.getByTestId("applyLink");
    expect(applyLink).toHaveAttribute("href", "apply");

    const positionTitle = screen.getByText("position");
    expect(positionTitle).toBeInTheDocument();
  });

  it("should hide title and description on mobile screen", () => {
    window.matchMedia = mockMatchMedia(true);

    render(<CompanyFooter apply="apply" position="position" />);

    const vacancyDescriptionTitle = screen.getByTestId(
      "vacancyDescription_title"
    );
    expect(vacancyDescriptionTitle).toHaveStyle("display: none");
  });

  test("renders all main text elements", () => {
    render(<VacancyDescription {...defaultProps} />);

    expect(screen.getByText(/3 days ago/i)).toBeInTheDocument();
    expect(screen.getByText(/Full-Time/i)).toBeInTheDocument();
    expect(screen.getByText(/Senior Developer/i)).toBeInTheDocument();
    expect(screen.getByText(/Remote/i)).toBeInTheDocument();

    expect(
      screen.getByText(/This is the job description/i)
    ).toBeInTheDocument();

    expect(
      screen.getByText(/You must have experience in React and Node.js./i)
    ).toBeInTheDocument();
    expect(screen.getByText(/TypeScript/i)).toBeInTheDocument();

    expect(
      screen.getByText(/You will work on our product team/i)
    ).toBeInTheDocument();
  });

  test("renders the apply button with correct behavior", () => {
    render(<VacancyDescription {...defaultProps} />);

    const applyButton = screen.getByText(/apply now/i);
    expect(applyButton).toBeInTheDocument();

    const applyLink = screen.getByRole("link", { name: /apply now/i });
    expect(applyLink).toHaveAttribute("href", "/apply");

    userEvent.hover(applyButton);
    userEvent.unhover(applyButton);
    userEvent.click(applyButton);
  });

  test("calls setFooterVisible when intersection changes", () => {
    const mockSetFooterVisible = jest.fn();
    render(
      <VacancyDescription
        {...defaultProps}
        setFooterVisible={mockSetFooterVisible}
      />
    );

    const observerInstance = (global.IntersectionObserver as jest.Mock).mock
      .calls[0][0];

    observerInstance([{ isIntersecting: false }]);
    expect(mockSetFooterVisible).toHaveBeenCalledWith(true);

    observerInstance([{ isIntersecting: true }]);
    expect(mockSetFooterVisible).toHaveBeenCalledWith(false);
  });
});
