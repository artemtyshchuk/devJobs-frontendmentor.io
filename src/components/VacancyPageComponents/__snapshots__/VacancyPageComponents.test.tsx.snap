// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VacancyPage should render Company Header correctly 1`] = `
<div
  class="companyHeader"
  data-testid="companyHeader"
>
  <div
    class="companyHeader_logoContainer"
  >
    <img
      alt="company logo"
      data-testid="companyLogo"
    />
  </div>
  <div
    class="companyHeader_companyInfoContainer"
  >
    <p
      class="companyHeader_companyName"
    >
      company
    </p>
    <p
      class="companyHeader_companyWebsite"
    >
      website
    </p>
  </div>
  <div
    class="companyHeader_companySiteContainer"
  >
    <a
      data-testid="companyWebsite"
      href="website"
      rel="noreferrer"
      target="_blank"
    >
      <button
        class="companyHeader_companySiteButton"
        data-testid="headerButton"
      >
        Company Site
      </button>
    </a>
  </div>
</div>
`;
